assemblies-valid/AssembliesValid.cs

```diff
@@ -0,0 +1,219 @@
+            if (exitCode != 0)
+            {
+                Console.Error.WriteLine("'dotnet' command not found");
+                Console.Error.WriteLine("PATH: " + Environment.GetEnvironmentVariable("PATH"));
+                Assert.True(false);
+            }
+            dotnetPath = dotnetPath.Trim();
+            exitCode = RunProcessAndGetOutput(new string[] { "readlink", "-f", dotnetPath }, out dotnetPath);
+            if (exitCode != 0)
+            {
+                Console.Error.WriteLine($"Unable to run readlink -f {dotnetPath}");
+                Assert.True(false);
+            }
+            dotnetPath = dotnetPath.Trim();
+
+            string searchRoot = new FileInfo(dotnetPath).DirectoryName;
+            var searchRootDirectory = new System.IO.DirectoryInfo(searchRoot);
+
+            Console.WriteLine($"Searching for dotnet binaries in {searchRoot}");
+
+            var machine = GetCurrentMachine();
+
+            ICollection<string> assemblies = FindAssemblyFiles(searchRootDirectory);
+
+            bool allOkay = true;
+            foreach (var assembly in assemblies)
+            {
+                bool inBlacklist = FileNameBlacklist.Any(basename => new FileInfo(assembly).Name == basename);
+
+                if (!inBlacklist) {
+                    using (var file = File.Open(assembly, FileMode.Open, FileAccess.Read))
+                    {
+                        var reader = new PEReader(file);
+                        bool hasAot = AssemblyHasAot(assembly, reader, machine);
+                        bool inReleaseMode = AssemblyIfNgenIsInReleaseMode(assembly, reader);
+
+                        if (hasAot && inReleaseMode)
+                        {
+                            Console.WriteLine($"{assembly}: OK");
+                        }
+                        else
+                        {
+                            Console.WriteLine($"error: {assembly} hasAot: {hasAot}, inReleaseMode: {inReleaseMode}");
+                            allOkay = false;
+                        }
+                    }
+                }
+            }
+
+            Assert.True(allOkay);
+        }
+
+        static int RunProcessAndGetOutput(string[] processAndArguments, out string standardOutput)
+        {
+            ProcessStartInfo startInfo = new ProcessStartInfo();
+            startInfo.FileName = processAndArguments[0];
+            startInfo.Arguments = string.Join(" ", processAndArguments.Skip(1));
+            startInfo.RedirectStandardOutput = true;
+
+            using (Process p = Process.Start(startInfo))
+            {
+                p.WaitForExit();
+                using (StreamReader r = p.StandardOutput)
+                {
+                    standardOutput = r.ReadToEnd();
+                }
+                return p.ExitCode;
+            }
+        }
+
+        static Machine GetCurrentMachine()
```

Comment:
At some point we may want to put these functions in a shared source folder, so multiple tests can use them.